numpy/core/src/multiarray/methods.c

```diff
@@ -619,10 +619,9 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
     if (fd == NULL) {
         goto fail;
     }
-    if (PyArray_ToFile(self, fd, sep, format) < 0) {
-        goto fail;
-    }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
+    res1 = PyArray_ToFile(self, fd, sep, format);
+    res2 = npy_PyFile_DupClose2(file, fd, orig_pos);
+    if (res1 < 0 || res2 < 0) {
         goto fail;
     }
     if (own && npy_PyFile_CloseFile(file) < 0) {
```

Comment:
if `res2 < 0`, you still should close this file

Reply:
Thanks @eric-wieser. I've updated it now so that it closes ```file``` even if ```npy_PyFile_DupClose2``` fails.